java/cuvs-java/src/main/java22/com/nvidia/cuvs/internal/BinaryQuantizerImpl.java

```diff
@@ -0,0 +1,175 @@
+import java.lang.foreign.Arena;
+import java.lang.foreign.MemorySegment;
+
+/**
+ * Implementation of binary quantizer functionality using Panama Foreign Function Interface.
+ */
+public class BinaryQuantizerImpl {
+
+  private static final long C_BYTE_SIZE = C_CHAR.byteSize();
+
+  /**
+   * Applies binary quantization transform to given dataset.
+   *
+   * @param cuvsResources CuVS resources
+   * @param dataset a two-dimensional float array to transform
+   * @return a two-dimensional byte array containing the binary quantized data
+   */
+  public static byte[][] transform(CuVSResources cuvsResources, float[][] dataset)
+      throws Throwable {
+    if (!(cuvsResources instanceof CuVSResourcesImpl)) {
+      throw new IllegalArgumentException("Unsupported " + cuvsResources);
+    }
+
+    CuVSResourcesImpl resources = (CuVSResourcesImpl) cuvsResources;
+
+    try (var localArena = Arena.ofConfined()) {
+      long rows = dataset.length;
+      long cols = rows > 0 ? dataset[0].length : 0;
+
+      MemorySegment datasetMemSegment = buildMemorySegment(localArena, dataset);
+      return performTransform(resources, localArena, rows, cols, datasetMemSegment, HOST_TO_DEVICE);
+    }
+  }
+
+  /**
+   * Applies binary quantization transform to given dataset.
+   *
+   * @param cuvsResources CuVS resources
+   * @param dataset a {@link Dataset} object containing the vectors to transform
+   * @return a two-dimensional byte array containing the binary quantized data
+   */
+  public static byte[][] transform(CuVSResources cuvsResources, Dataset dataset) throws Throwable {
+    if (!(cuvsResources instanceof CuVSResourcesImpl)) {
+      throw new IllegalArgumentException("Unsupported " + cuvsResources);
+    }
+
+    CuVSResourcesImpl resources = (CuVSResourcesImpl) cuvsResources;
+
+    try (var localArena = Arena.ofConfined()) {
+      long rows = dataset.size();
+      long cols = dataset.dimensions();
+
+      MemorySegment datasetMemSegment = ((DatasetImpl) dataset).seg;
+      return performTransform(
+          resources, localArena, rows, cols, datasetMemSegment, INFER_DIRECTION);
+    }
+  }
+
+  /**
+   * Core transformation logic shared by both overloads.
+   */
+  private static byte[][] performTransform(
+      CuVSResourcesImpl resources,
+      Arena localArena,
+      long rows,
+      long cols,
+      MemorySegment datasetMemSegment,
+      Util.CudaMemcpyKind memcpyDirection)
+      throws Throwable {
+
+    long cuvsResourcesPtr = resources.getMemorySegment().get(cuvsResources_t, 0);
+
+    // Allocate device memory for input and output
+    MemorySegment datasetD = localArena.allocate(C_POINTER);
+    MemorySegment outputD = localArena.allocate(C_POINTER);
+
+    long datasetBytes = C_FLOAT_BYTE_SIZE * rows * cols;
+    long outputCols = (cols + 7) / 8; // Ceiling division for bit packing
+    long outputBytes = C_BYTE_SIZE * rows * outputCols;
+
+    int returnValue = cuvsRMMAlloc(cuvsResourcesPtr, datasetD, datasetBytes);
```

Comment:
Similarly, https://github.com/rapidsai/cuvs/pull/1069 introduces an `allocateRMMSegment` utility to replace the need to allocate + get (but that has not been merged yet).

Reply:
Please be aware that I'm considering moving `allocateRMMSegment` from `Util` to `CloseableRMMAllocation` in #1215.  There might be some minor changes in this bit.